### PR TITLE
8327741: JVM crash in hotspot/share/opto/compile.cpp - failed: missing inlining msg

### PR DESCRIPTION
**Pull request by rwestrel**
The crash occurs when a virtual call is devirtualized late. Inlining
is not attempted then. So no new inlining diagnostic message is
produced which causes the assert failure. There's some valuable
information that can be reported though (the call is
devirtualized).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8327741](https://bugs.openjdk.org/browse/JDK-8327741): JVM crash in hotspot/share/opto/compile.cpp - failed: missing inlining msg (**Bug** - P4)

### Reviewers
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18685/head:pull/18685` \
`$ git checkout pull/18685`

Update a local copy of the PR: \
`$ git checkout pull/18685` \
`$ git pull https://git.openjdk.org/jdk.git pull/18685/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18685`

View PR using the GUI difftool: \
`$ git pr show -t 18685`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18685.diff">https://git.openjdk.org/jdk/pull/18685.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18685#issuecomment-2044607540)

### CODE REVIEW

**Review by iwanowww — COMMENTED** [review]
Looks good. Thanks for fixing it.

I'll submit it for testing.

File: test/hotspot/jtreg/compiler/print/TestPrintInliningLateVirtualCall.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8327741
+ * @summary JVM crash in hotspot/share/opto/compile.cpp - failed: missing inlining msg
+ * @run main/othervm -XX:-BackgroundCompilation -XX:+PrintCompilation -XX:+PrintInlining TestPrintInliningLateVirtualCall
```
Comment (iwanowww) [inline test/hotspot/jtreg/compiler/print/TestPrintInliningLateVirtualCall.java:28]: The test misses `-XX:+UnlockDiagnosticVMOptions` flag:
```
Error: VM option 'PrintInlining' is diagnostic and must be enabled via -XX:+UnlockDiagnosticVMOptions.
Error: The unlock option must precede 'PrintInlining'.
```

Comment (rwestrel) [reply test/hotspot/jtreg/compiler/print/TestPrintInliningLateVirtualCall.java]: Right. Fixed in new commit.

**Review by iwanowww — APPROVED** [review]
Looks good.

**Review by chhagedorn — APPROVED** [review]
Looks good!

**Comment by rwestrel** [comment]
@iwanowww @chhagedorn thanks for the reviews

**Pull request closed without merging**